packages/did/src/Did.chain.ts

```diff
@@ -105,48 +149,119 @@ function decodeDidChainRecord(
   return didRecord
 }
 
+// Same reasoning as `assembleKeyId`.
+function decodeServiceChainRecord(
+  serviceDetails: IServiceEndpointChainRecordCodec,
+  did: IDidDetails['did']
+): IDidServiceEndpoint {
+  const decodedId = hexToString(serviceDetails.id.toString())
+  return {
+    id: `${did}#${decodedId}`,
+    types: serviceDetails.serviceTypes.map((type) =>
+      hexToString(type.toString())
+    ),
+    urls: serviceDetails.urls.map((url) => hexToString(url.toString())),
+  }
+}
+
 export async function queryById(
   didIdentifier: IIdentity['address']
 ): Promise<IDidChainRecordJSON | null> {
-  const result = await queryEncoded(didIdentifier)
-  if (result.isSome) {
-    return decodeDidChainRecord(
-      result.unwrap(),
-      getKiltDidFromIdentifier(didIdentifier, 'full')
-    )
+  const result = await queryDidEncoded(didIdentifier)
+  if (result.isNone) {
+    return null
   }
-  return null
+  return decodeDidChainRecord(
+    result.unwrap(),
+    getKiltDidFromIdentifier(didIdentifier, 'full')
+  )
 }
 
-export async function queryByDID(
-  did: IDidChainRecordJSON['did']
+// Query full DID details given the DID URI.
+export async function queryDidDetails(
+  didUri: IDidDetails['did']
 ): Promise<IDidChainRecordJSON | null> {
-  // we will have to extract the id part from the did string
-  const didId = getIdentifierFromKiltDid(did)
-  return queryById(didId)
+  const { identifier, fragment } = parseDidUrl(didUri)
+  if (fragment) {
+    throw new Error(`The provided URI ${didUri} must not contain any fragment.`)
+  }
+  return queryById(identifier)
 }
 
-export async function queryKey(
-  did: string,
-  keyId: string
+// Query a given key given the DID identifier and the key ID.
+export async function queryDidKey(
+  keyUri: IDidKeyDetails['id']
 ): Promise<IDidKeyDetails | null> {
-  const encoded = await queryEncoded(getIdentifierFromKiltDid(did))
-  if (encoded.isNone) return null
-  const keyIdU8a = Crypto.coToUInt8(keyId)
-  let key: IDidKeyDetails | null = null
-  encoded.unwrap().publicKeys.forEach((keyDetails, id) => {
-    if (id.eq(keyIdU8a)) {
-      key = decodeDidPublicKeyDetails(did, id, keyDetails)
-    }
-  })
-  return key
+  const { identifier, fragment } = parseDidUrl(keyUri)
+  if (!fragment) {
+    throw new Error(
+      `The provided URI ${keyUri} does not contain a valid fragment for key ID.`
+    )
+  }
+  const didDetails = await queryById(identifier)
+  if (!didDetails) {
+    return null
+  }
+  return (
```

Comment:
Would be cleaner if you filter first and then return the result.

Reply:
I guess this is a matter of personal preference. I always prefer to exclude (and early return) when some condition fails so that we don't need any nesting or any return after a long block.